### ISSUE TITLE
AttributeError when passing tf.Variable into tf.function 

### ISSUE THREAD

**Issue by dovahcrow** [tensorflow/tensorflow#32748]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: no
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): v2.0.0-rc2
- Python version: 3.7.2
- Bazel version (if compiling from source): 0.25.2
- GCC/Compiler version (if compiling from source): 5.4.0 20160609
- CUDA/cuDNN version: None
- GPU model and memory: None

**Describe the current behavior**
```python
import tensorflow as tf
@tf.function
def foo(v):
    pass

foo(tf.Variable(1))
foo(tf.Variable(1))
```

Using this code piece, tensorflow throws an error: `AttributeError: 'NoneType' object has no attribute 'shape'`

**Describe the expected behavior**
No error occured.

**Code to reproduce the issue**
Provided above

**Other info / logs**
None

**Comment by gadagashwini** [comment]
@dovahcrow,
Workaround can be change any one of `tf.Variable` to `tf.constant` or both to `tf.constant`.
```
import tensorflow as tf
@tf.function
def foo(v):
    pass

foo(tf.constant(1))
foo(tf.constant(1))
```
Let us know this helps. Thanks!

**Comment by dovahcrow** [comment]
@gadagashwini Thanks! I currently solve this by using `tf.identity` to convert the `Variable` to a `Tensor`. But still, I wonder, is this a bug or feature?

**Comment by gadagashwini** [comment]
@dovahcrow,
I could reproduce the issue with TF 2.0.0.rc2. Please see the gist [here](https://colab.sandbox.google.com/gist/gadagashwini/4960b9907cbf83daae6a643136653125/untitled164.ipynb). Thanks!

**Comment by jvishnuvardhan** [comment]
@dovahcrow @tf.function on any defined function allows it to retrace the function when necessary to generate the correct graphs. 

I think your intention is not to control and trace the graph. In that case, you could use the following code

```
import tensorflow as tf
@tf.function
def foo(v):
    pass
a = tf.Variable(1)
b = tf.Variable(1)
foo(a)
foo(b)
```

Please check the resource [here](https://www.tensorflow.org/beta/tutorials/eager/tf_function#when_to_retrace) for more details on the retrace.

**Comment by dovahcrow** [comment]
@jvishnuvardhan Hi, I don't think this is because of the misuse of the API.
First, by changing the code piece to 
```python
import tensorflow as tf
@tf.function
def foo(v):
    pass
a = tf.Variable(1)
foo(a)
a = tf.Variable(1)
foo(a)
```
Tensorflow will raise the error I reported. In this case Tensorflow is disallowing my redefining the variable which is totally a legal operation in Python. This is very unexpected and super hard to debug when happened in a large codebase (in my case I spent 5 hours to reduce thousands lines of code to such a simple case.)
Second, as per the document said:
> For all other Python types, the keys are based on the object id() so that methods are traced independently for each instance of a class. In the future, TensorFlow may add more sophisticated caching for Python objects that can be safely converted to tensors.
```python
import tensorflow as tf
@tf.function
def foo(v):
    pass
a = tf.Variable(1)
print(id(a))
foo(a)
a = tf.Variable(1)
print(id(a))
foo(a)
```
I get this as the output
```
140522833653432
140522833649960
```
, as well as the error
```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/.pyenv/versions/3.7.2/lib/python3.7/site-packages/tensorflow_core/python/eager/function.py in _hash_fix(self, elem)
    108     try:
--> 109       hash(elem)
    110     except TypeError:

TypeError: weak object has gone away
```
This indicates `@tf.function` somehow didn't retrace the graph even if the `id`s of the objects are different.

**Comment by ShahinMojarad** [comment]
Thank you guys for the notes. I spend two days to realize that when using @tf.function causes the error and when removing @tf.function everything works fine. Then I came up with search phrase of "AttributeError: 'NoneType' object has no attribute 'shape' + when using tf.function" and I found this post. I fixed the issue by using tf.constant when feeding training set to the function. I hope they fix it soon, it's very hard to troubleshoot this misleading AttributeError.
